### PR TITLE
SparseSwap

### PR DESCRIPTION
**Pull request by yugure-orca**
# Problems
## Uninitialised TA problem
- The absence of TickArray causes the trade to stop
- It is difficult to explain the problems caused by uninitialised TickArrays to users.
- It is impractical to initialise a large number of TickArrays in a FullRange pool
- It may be not sustainable to initialize many tickarrays by protocol (~ Orca)

## Problems with the order of TickArrays passed to instructions
It is necessary to provide a TickArray containing `tick_current_index` as `tick_array_0`.
However, if there is a current price near the boundary of the TickArray, this condition may not be met because the price moves; it is a kind of slippage, but even if the slippage is increased, an error occurs in this case.

# Solution
## Accept uninitialized TickArray
swap, swapV2, twoHopSwap, twoHopSwapV2 will accept uninitialized TickArray account address.

These instructions will treat uninitialized TickArray account as if they are zero padded and process trade.

## On-chain TickArray account order adjustment
swap, swapV2, twoHopSwap, twoHopSwapV2 will accept (maybe uninitialized) TickArray account address in any order and sort them as needed.  If instruction receive required TickArray accounts, it should work.

## Additional TickArray accounts
swapV2 and twoHopSwapV2 will accept additional up to 3 TickArray accounts via remaining_accounts.

- This is V2 only because V1 doesn't use remaining_accounts.
- The maximum number of accounts received will be six, but the maximum number of TickArrays used in a single trade will remain the same at three. This means that we can provide a backup TickArray in case prices move." The name "supplemental" comes from this use.

# Implementation Notes
## No breaking change on instruction interface
- There will be no change in the accounts received by v1.
- There are additions to the remaining_accounts that v2 will receive, but these do not affect the current program.

## Switching AccountLoader to UncheckedAccount
AccountLoader can only accept accounts that have been initialized (or at least the owner program has been changed to initialize account).
Therefore, to receive uninitialized accounts, use UncheckedAccount.

The verification that was conventionally done via has_one needs to be done via handler.
Also, the verification done internally by load_mut() needs to be done the same way.

## ProxiedTickArray
Add one abstraction layer, ProxiedTickArray, to minimize changes to the current battle-tested `SwapTickSequence` code.

For TickArrays that have been initialized, the process is delegated to the existing implementation. For uninitialized TickArray, the process is delegated to ZeroedTickArray where all Tick are zero initialized.

### Notes: Heap size limination
Initially, we considered placing the entire 0-initialized TickArray account in the Heap and using RefMut to that account, but the Heap size limit is 32 KB, so we were limited to two accounts at most.

Since twoHopSwapV2 handles a maximum of six uninitialized accounts, this method could not be used.

## On-chain TickArray account order adjustment
First, calculate up to three start_tick_indexes for the required TickArray and verify if a TickArray account with that start_tick_index is provided.

With this method, if an account is provided that has been initialized as it is now, no PDA calculation or pubkey comparison occurs, since matching is possible only by comparing the start_tick_indexes among themselves.

The PDA calculation is only performed if no account with start_tick_index is found among the initialized accounts.

Uninitialized accounts are verified as uninitialized.
The conditions for uninitialization are that SystemProgram is the owner and the data size is 0.

Whether the uninitialized account is a correct TickArray address or not is not verified because it is impossible to verify (we need to find seend for them). However, it is used only if it matches the correct PDA derived using start_tick_index.  So wrong accounts will be simply ignored.

### CODE REVIEW

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 1]
```diff
@@ -0,0 +1,1746 @@
...
+            }
+        })
+        .collect::<Vec<i32>>();
+
+    start_tick_indexes
+}
+
+fn is_valid_start_tick_index(start_tick_index: i32, ticks_in_array: i32) -> bool {
+    // caller must assure that start_tick_index is multiple of ticks_in_array
+    start_tick_index + ticks_in_array > MIN_TICK_INDEX && start_tick_index < MAX_TICK_INDEX
```
Comment (philcchen) [inline programs/whirlpool/src/util/sparse_swap.rs:348]: hmm, I think there is a discrepancy between this and the existing `check_is_valid_start_tick`, I believe the existing `check_is_valid_start_tick` uses an equality condition on the`MIN_TICK_INDEX`, i.e. `start_tick_index + ticks_in_array >= MIN_TICK_INDEX`?

**Review by philcchen — COMMENTED** [review]
Took a pass over the core logic code, will do another pass over edge cases and tests

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 2]
```diff
@@ -0,0 +1,1746 @@
...
+    }
+
+    pub fn build<'a>(&'a self) -> Result<SwapTickSequence<'a>> {
+        let mut tick_array_refmuts = VecDeque::with_capacity(3);
+        for tick_array_account in self.tick_array_accounts.iter() {
+            match tick_array_account {
+                TickArrayAccount::Initialized { account_info, .. } => {
+                    use std::ops::DerefMut;
+
+                    let data = account_info.try_borrow_mut_data()?;
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:241]: good point!

In my understanding, it is just memory mapping operation (pointer operation in low-level), so no significant cost.
And I think we've already done this processing (load & load_mut).
https://github.com/orca-so/whirlpools/blob/main/programs/whirlpool/src/instructions/two_hop_swap.rs#L44

But I agree that it is nice to merge `try_from` and `build` and dedup this logic.

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 3]
```diff
@@ -0,0 +1,1746 @@
...
+        if tick_array_accounts.is_empty() {
+            return Err(crate::errors::ErrorCode::InvalidTickArraySequence.into());
+        }
+
+        Ok(Self {
+            tick_array_accounts,
+        })
+    }
+
+    pub fn build<'a>(&'a self) -> Result<SwapTickSequence<'a>> {
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:234]: Initially, a dummy TickArray was allocated in the heap; it was necessary to maintain ownership of the heap in order to use RefMut.
The Builder was more significant for not cluttering up the handler with variables for this heap than for its logic.
However, ZeroedTickArray probably doesn't need to do this anymore.

will try to refactor.

File: programs/whirlpool/src/util/sparse_swap.rs (hunk 1 above)
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:348]: Good catch!
will check this edge case.

Good news is that 443636 (abs value of MIN_TICK_INDEX) is not multileple of 88.
So this MIN_TICK_INDEX will never be start_tick_index. 😁

File: programs/whirlpool/src/state/tick.rs [hunk 4]
```diff
@@ -297,25 +297,19 @@ impl TickArray {
...
+        in_search_range(self.start_tick_index, tick_index, tick_spacing, shifted)
     }
 
     pub fn check_in_array_bounds(&self, tick_index: i32, tick_spacing: u16) -> bool {
         self.in_search_range(tick_index, tick_spacing, false)
     }
 
     pub fn is_min_tick_array(&self) -> bool {
-        self.start_tick_index <= MIN_TICK_INDEX
+        is_min_tick_array(self.start_tick_index)
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/state/tick.rs]: cool! 🚀
updated with `TickArrayType` trait.

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 5]
```diff
@@ -0,0 +1,1746 @@
...
+    pub fn new_initialized(refmut: RefMut<'a, TickArray>) -> Self {
+        ProxiedTickArray::Initialized(refmut)
+    }
+
+    pub fn new_uninitialized(start_tick_index: i32) -> Self {
+        ProxiedTickArray::Uninitialized(ZeroedTickArray::new(start_tick_index))
+    }
+
+    pub fn start_tick_index(&self) -> i32 {
+        match self {
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs]: cool again 👍
updated with AsRef & AsMut.

File: programs/whirlpool/src/util/sparse_swap.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:229]: I tried for a while to merge `try_from` and `build`, but the lifetime issue is troublesome, so I would like to leave it as it is.

File: programs/whirlpool/src/util/sparse_swap.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:222]: I tried for a while to merge `try_from` and `build`, but the lifetime issue is troublesome, so I would like to leave it as it is.

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 6]
```diff
@@ -0,0 +1,1746 @@
...
+pub struct SparseSwapTickSequenceBuilder<'info> {
+    tick_array_accounts: Vec<TickArrayAccount<'info>>,
+}
+
+impl<'info> SparseSwapTickSequenceBuilder<'info> {
+    pub fn try_from(
+        whirlpool: &Account<'info, Whirlpool>,
+        a_to_b: bool,
+        static_tick_array_account_infos: Vec<AccountInfo<'info>>,
+        supplemental_tick_array_account_infos: Option<Vec<AccountInfo<'info>>>,
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:130]: Agreed.
Added comments.

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 7]
```diff
@@ -0,0 +1,1746 @@
...
+            tick_array_refmuts.pop_front(),
+            tick_array_refmuts.pop_front(),
+        ))
+    }
+}
+
+fn peek_tick_array<'info>(account_info: AccountInfo<'info>) -> Result<TickArrayAccount<'info>> {
+    use anchor_lang::Discriminator;
+
+    // following process is ported from anchor-lang's AccountLoader::try_from and AccountLoader::load_mut
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:254]: It is difficult to satisfy the reference condition required by AccountLoader::try_from.
Added comment as to why.

Probably related issue: https://github.com/coral-xyz/anchor/pull/2770

File: programs/whirlpool/src/util/sparse_swap.rs (hunk 1 above)
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs]: I chose the simplest approach.
Use `Tick::check_is_valid_start_tick` 😁

File: programs/whirlpool/src/util/sparse_swap.rs [hunk 8]
```diff
@@ -0,0 +1,1746 @@
...
+                    //   - Data size is zero
+                    //   - Writable account
+                    // But we are not sure if these accounts are valid TickArray PDA for this whirlpool,
+                    // so we need to check it later.
+                    uninitialized.push((*account_address, state));
+                }
+            }
+        }
+
+        let start_tick_indexes = get_start_tick_indexes(whirlpool, a_to_b);
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/sparse_swap.rs:180]: Added comment.👍

PDA calculation is a high cost process that costs 3000 CU/PDA ~, so if it can be matched against start_tick_index, it is preferred.

**Review by philcchen — APPROVED** [review]

**Comment by yugure-orca** [comment]
merged FROM the latest main to resolve conflict on errors.rs.
Ready for the audit.

**Comment by yugure-orca** [comment]
The audit for [3f30966](https://github.com/orca-so/whirlpools/pull/159/commits/3f309667c3773bebb143566f69d50b1575ca7c18) have been completed.
